### PR TITLE
[v1][sampler] Inplace logprobs comparison to get the token rank

### PR DESCRIPTION
**Pull request by houseroad**
Summary: Original implementation is inefficient, since we will create a copy of the original tensor. So create a triton kernel to get this done in place.

Test Plan

Launch the server with warm up requests

Before:
OOM

After:
No more OOM

Differential Revision: D78606604

### CODE REVIEW

File: vllm/v1/sample/ops/logprobs.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    # 4. Create a mask to avoid reading out of bounds.
+    mask = offsets < n_elements
+
+    # 5. Load a block of data from the correct row of `x`.
+    # We use the mask to safely handle rows whose size is not a multiple of BLOCK_SIZE.
+    x_block = tl.load(x_row_start_ptr + offsets * x_stride_elem, mask=mask)
+
+    # 6. Perform the comparison and sum the result within this block.
+    # `x_block > value` creates a boolean vector, and tl.sum treats True as 1.
+    local_count = tl.sum(x_block > value)
```
Comment (Jialin) [inline vllm/v1/sample/ops/logprobs.py]: IIUC, this PR is for HBM usage reduction?
- Originally, x >= values would create a new bool tensor before count True against it, so the HBM overhead would be x.nbytes.
- In the new implementation, x_block > value would still create boolean vector before accumulatiing with tl.sum. However, blocks are not executed at the exact same time, so the HBM usage would become max_concurrent_block * x_block.nbytes

Is my understand accurate?

I kinda feel we're having quite a lot usage similar to (x >= y).sum(), should we introduce a new pytorch API to optimize this, in general.

**Comment by Jialin** [comment]
Maybe add unittest to confirm the correctness?

File: vllm/v1/sample/sampler.py [hunk 2]
```diff
@@ -174,7 +176,7 @@ def gather_logprobs(
         token_logprobs = logprobs.gather(-1, token_ids)
 
         # Compute the ranks of the actual token.
-        token_ranks = (logprobs >= token_logprobs).sum(-1)
```
Comment (yinghai) [inline vllm/v1/sample/sampler.py:177]: in general I think the logit processor is quite inefficient with a bunch of intermediates. I think adopting some compilation techniques e.g. https://github.com/flashinfer-ai/flashinfer/blob/main/flashinfer/logits_processor/compiler.py would be good. I also think this is actually a good use case for torch.compile.

**Comment by houseroad** [comment]
Actually, switched to torch.compile, no more OOM. Thanks to @yinghai for the suggestion.

**Review by zou3519 — APPROVED** [review]

**Comment by yeqcharlotte** [comment]
@houseroad @zou3519 will this be JIT and trigger compilation on the fly?

**Comment by houseroad** [comment]
The overhead should be fine, since it's only a few ops, usually this is what torch.compile is really great at.

**Comment by zou3519** [comment]
This will trigger compilation on the fly. It should only compile once.

It is possible to enforce this via torch.compiler.set_stance("fail_on_recompiles") and/or vLLM's existing support_torch_compile mechanism. vLLM's existing mechanism might be too heavy for this, but in the long term we may want something that interacts better with the vLLM-compile cache.

**Comment by yeqcharlotte** [comment]
> This will trigger compilation on the fly. It should only compile once.
> 
> It is possible to enforce this via torch.compiler.set_stance("fail_on_recompiles") and/or vLLM's existing support_torch_compile mechanism. vLLM's existing mechanism might be too heavy for this, but in the long term we may want something that interacts better with the vLLM-compile cache.

I see. Essentially we may need some proper warm-up to avoid weird latency spikes?

**Comment by yinghai** [comment]
I think you can do static shape compilation for this and I think vLLM does more than enough warmup runs lol. It specifically does a warmup with max_request sampling path being triggered.
